internal/provider/collection/data_source.go

```diff
@@ -0,0 +1,139 @@
+package collection
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	v20230301 "github.com/smallstep/terraform-provider-smallstep/internal/apiclient/v20230301"
+	"github.com/smallstep/terraform-provider-smallstep/internal/provider/utils"
+)
+
+var _ datasource.DataSourceWithConfigure = (*DataSource)(nil)
+
+func NewDataSource() datasource.DataSource {
+	return &DataSource{}
+}
+
+// DataSource implements data.smallstep_collection
+type DataSource struct {
+	client *v20230301.Client
+}
+
+func (a *DataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = collectionTypeName
+}
+
+// Configure adds the Smallstep API client to the data source.
+func (a *DataSource) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*v20230301.Client)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Get Smallstep API client from provider",
+			fmt.Sprintf("Expected *v20230301.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+
+	a.client = client
+}
+
+func (a *DataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var config Model
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &config)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	httpResp, err := a.client.GetCollection(ctx, config.Slug.ValueString(), &v20230301.GetCollectionParams{})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Smallstep API Client Error",
+			fmt.Sprintf("Failed to read collection %s: %v", config.Slug.String(), err),
+		)
+		return
+	}
+	defer httpResp.Body.Close()
+
+	if httpResp.StatusCode != http.StatusOK {
+		resp.Diagnostics.AddError(
+			"Smallstep API Response Error",
+			fmt.Sprintf("Received status %d reading collection %s: %s", httpResp.StatusCode, config.Slug.String(), utils.APIErrorMsg(httpResp.Body)),
+		)
+		return
+	}
+
+	collection := &v20230301.Collection{}
+	if err := json.NewDecoder(httpResp.Body).Decode(collection); err != nil {
+		resp.Diagnostics.AddError(
+			"Smallstep API Client Error",
+			fmt.Sprintf("Failed to unmarshal collection %s: %v", config.Slug.String(), err),
+		)
+		return
+	}
+
+	remote, d := fromAPI(ctx, collection, req.Config)
+	if d.HasError() {
+		resp.Diagnostics.Append(d...)
```

Comment:
Any reason to conditionally append the diagnostics? `fromAPI` for collection doesn't produce any diagnostics right now, but in other places (eg: provisioner data source) it just assumes to append all diagnotics regardless if an error was emitted.

Reply:
0ac66ff